src/main/kotlin/org/codefreak/codefreak/service/file/FileSystemFileService.kt

```diff
@@ -0,0 +1,438 @@
+
+    if (Files.exists(target)) {
+      require(
+        Files.isDirectory(target) || arePathsEqual(source, target)
+      ) { "The target path `$target` already exists" }
+    }
+
+    var targetPath = target
+
+    if (Files.isDirectory(target)) {
+      targetPath = Paths.get(target.toString(), source.fileName.toString())
+    }
+
+    Files.move(source, targetPath)
+  }
+
+  private fun isDescendant(path: Path, of: Path): Boolean {
+    return path.toString().startsWith(of.toString())
+  }
+
+  private fun isBasenamePresent(source: Path, target: Path): Boolean {
+    if (Files.isRegularFile(source)) {
+      val path = Paths.get(target.toString(), source.fileName.toString())
+      return Files.exists(path) && !arePathsEqual(source, target)
+    }
+
+    return Files.list(source).anyMatch {
+      isBasenamePresent(it, target)
+    }
+  }
+
+  private fun arePathsEqual(path1: Path, path2: Path): Boolean {
+    return path1.toString() == path2.toString() ||
+        path1.parent.toString() == path2.toString() ||
+        path1.toString() == path2.parent.toString()
+  }
+
+  private fun moveDirectoryRecursively(source: Path, target: Path) {
+    require(Files.isDirectory(source)) { "`$source` is not a directory" }
+
+    source.let {
+      val targetPath = Paths.get(target.toString(), it.fileName.toString())
+      Files.createDirectories(targetPath)
+    }
+
+    Files.walkFileTree(source, object : SimpleFileVisitor<Path>() {
+      override fun preVisitDirectory(dir: Path?, attrs: BasicFileAttributes?): FileVisitResult {
+        dir?.let {
+          // Create the directory in the target so its files can be copied there
+          val sourceDir = source.relativize(it).normalize()
+          val targetDir = Paths.get(target.toString(), sourceDir.toString())
+          Files.createDirectories(targetDir)
+        }
+        return FileVisitResult.CONTINUE
+      }
+
+      override fun visitFile(file: Path?, attrs: BasicFileAttributes?): FileVisitResult {
+        file?.let {
+          // Move the file to the new location
+          val sourcePath = source.relativize(it).normalize()
+          val targetPath = Paths.get(target.toString(), sourcePath.toString())
+          moveSingleFile(it, targetPath)
+        }
+        return FileVisitResult.CONTINUE
+      }
+
+      override fun postVisitDirectory(dir: Path?, exc: IOException?): FileVisitResult {
+        dir?.let {
+          // Delete directory after moving all its contents
+          Files.delete(it)
+        }
+        return FileVisitResult.CONTINUE
+      }
+    })
+  }
+
+  override fun writeFile(collectionId: UUID, path: String): OutputStream {
+    val collectionPath = getCollectionPath(collectionId)
+    val filePath = Paths.get(collectionPath.toString(), path)
+
+    require(!Files.isDirectory(filePath)) { "`$path` is a directory" }
+    require(isValidPath(collectionId, filePath)) { "`$path` is not a valid path" }
+
+    return Files.newOutputStream(filePath)
```

Comment:
Maybe we should note that the default behavior is `CREATE`, `TRUNCATE_EXISTING`, and `WRITE`

Reply:
Umm.. how?

Reply:
As a comment 😁